Use CHaP :robot:

ch1bo opened the pull request:
:robot: The Cardano Haskell Packages (CHaP) will make our life's simpler by forcing our upstream dependencies to version & release their code, and we don't need to keep `source-repository-package` entries in the `cabal.project`

:robot: Added the CHaP index to `cabal.project` and made it work using nix. The nix shell will now compile (or download) & provide most third-party dependencies to us.

:robot: Only exception: `cardano-api` where only version `1.35.3` is in CHaP, which does not include our upstreamed changes. Needed to backport the two commits and keep the `source-repository-package` stanza for that one.

ch1bo committed:
Use the cardano-haskell-packages cabal index

This will make dependency management (hopefully) a lot simpler and
improve UX with cabal and nix(haskell.nix)

ch1bo committed:
Make CHaP available via nix as well

This will result in quite some rebuilds, but our nix cache is now
containing ALL the cardano haskell packages :tada:

ch1bo committed:
Use quickcheck-dynamic from hackage

ch1bo committed:
Add contributing guidelines about updating dependencies

v0d1ch reviewed (approved):
Nice! Builds on my machine :muscle:

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -349,7 +349,7 @@ test-suite tests
     , process
     , QuickCheck
-    , quickcheck-dynamic
+    , quickcheck-dynamic                <2.0.0
```
Yes. This was automatic by `cabal-fmt`

ch1bo committed:
Don't use pure nix-shell in CI

For some reason we need to fetch the CHaP index and this does not play
well with pure nix-shells. Alternative would've been to add TLS
capability to the nix-shell .. I guess. But this is not important enough
that I would bother now.
